File: android/app/src/main/java/com/TMR24/MotionStudy/UserShiftHistory.java
package com.TMR24.MotionStudy;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.firebase.Timestamp;
import com.google.firebase.firestore.FirebaseFirestore;
import com.google.firebase.firestore.QueryDocumentSnapshot;
import com.google.firebase.firestore.QuerySnapshot;
import com.google.firebase.functions.FirebaseFunctions;

import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map;

public class UserShiftHistory extends AppCompatActivity {
    private FirebaseFirestore db;
    private String  TAG, userNameEmail, parentHierarchyPositionUser;
    private String  nameOrganization, nameSubdivision, namePosition,  idPosition, idOrganization, idSubdivision, idDocPositionUser, userСomment;
    private TextView textActivPositionHistori;
    private Button  buttonToReturnHistori;
    private ListView listViewHistoriButton;
    private ArrayAdapter<String> adapterHistoriButton;
    private List <String> listHistoriButton, listHistoriButtonItem;
    private Timestamp workShiftEndTime, workShiftStartTime;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_user_shift_history);
        init();
        setOnClickItemlistViewHistoriButton();
    }
    private void init()
    {
        db = FirebaseFirestore.getInstance();
     //   mFunctions = FirebaseFunctions.getInstance();

        buttonToReturnHistori = findViewById(R.id.buttonToReturnHistori);
        listViewHistoriButton = findViewById(R.id.listViewHistori);
        listHistoriButton = new ArrayList <>();
        listHistoriButtonItem = new ArrayList<>();
        adapterHistoriButton = new ArrayAdapter <>(this, android.R.layout.simple_list_item_1, listHistoriButton);
        listViewHistoriButton.setAdapter(adapterHistoriButton);

    }
    public void onStart() {

        super.onStart();
        Intent i = getIntent();
        if (i != null)
        {   //очистили ArrayList
            listHistoriButton.clear();
            listHistoriButtonItem.clear();
            // получили строку данных с предидущего Активити
            userNameEmail = i.getStringExtra(Constant.USER_NAME_EMAIL);
            parentHierarchyPositionUser = i.getStringExtra(Constant.PARENT_HIERARCHY_POSITION_USER);
            String delimeter = ">";
            idOrganization = parentHierarchyPositionUser.split(delimeter)[0];
            nameOrganization = parentHierarchyPositionUser.split(delimeter)[1];
            idSubdivision = parentHierarchyPositionUser.split(delimeter)[2];
            nameSubdivision = parentHierarchyPositionUser.split(delimeter)[3];
            idPosition = parentHierarchyPositionUser.split(delimeter)[4];
            namePosition = parentHierarchyPositionUser.split(delimeter)[5];
            idDocPositionUser = parentHierarchyPositionUser.split(delimeter)[6];
            userСomment = parentHierarchyPositionUser.split(delimeter)[7];
            //вывели на экран Должность Подразделение Организацию в которой планируем работать
            TextView textActivPosition = findViewById(R.id.textActivPositionHistori);
            textActivPosition.setText(nameOrganization+" > "+nameSubdivision+" > "+namePosition);
            //проверили нет ли активной смены
            db.collection("WorkShift")
                    .whereEqualTo("EmailPositionUser", userNameEmail)
                    .whereEqualTo("WorkShiftEnd", "false")
                    .whereEqualTo("IdDocPosition", idPosition)
                    .get()
                    .addOnCompleteListener(new OnCompleteListener < QuerySnapshot >()
                    {
                        @Override
                        public void onComplete(@NonNull Task <QuerySnapshot> task)
                        {
                            if (task.isSuccessful()) {
                                for (QueryDocumentSnapshot document : task.getResult()) {
                                    //Имеется активная смена
                                    Log.d(TAG, document.getId() + " => " + document.getData());
                                    Map <String, Object> doc = document.getData();
                                    workShiftStartTime = (Timestamp) doc.get("WorkShiftStartTime");
                                    Date g = workShiftStartTime.toDate();
                                    long x = workShiftStartTime.getSeconds();
                                    SimpleDateFormat formatForDateNow = new SimpleDateFormat("dd.MM.yyyy ' time ' k:mm:ss");
                                    workShiftEndTime = (Timestamp) doc.get("WorkShiftEndTime");
                                    Date h = workShiftEndTime.toDate();
                                    long a = workShiftEndTime.getSeconds();
                                    long seconds = a-x;
                                    // переводим секунды в дни-часы-минуты-секунды
                                    // long sec = seconds % 60;
                                    long minutes = seconds % 3600 / 60;
                                    long hours = seconds % 86400 / 3600;
                                    long days = seconds / 86400;
                                    listHistoriButton.add("Shift start "+formatForDateNow.format(g)+" > End of shift "+formatForDateNow.format(h));
                                    listHistoriButtonItem.add(document.getId()+">"+"Day " + days + " Hour " + hours + " Minute " + minutes);
                                    adapterHistoriButton.notifyDataSetChanged();
                                }
                            } else {
                                //отсутствует Активная смена
                                Log.d(TAG, "Error getting documents: ", task.getException());

                            }
                        }
                    });
        }
    }
    private void setOnClickItemlistViewHistoriButton()
    { //
        listViewHistoriButton.setOnItemClickListener(new AdapterView.OnItemClickListener()
        {
            @Override
            public void onItemClick(AdapterView < ? > parent, View view, int position, long id)
            {
                String buttonHistori = listHistoriButtonItem.get(position);
                String delimeter = ">";
                String idButtonHistori = buttonHistori.split(delimeter)[0];
                String shiftDurationHistori = buttonHistori.split(delimeter)[1];
                // выводим диалоговое окно
                AlertDialog.Builder builder = new AlertDialog.Builder(UserShiftHistory.this);
                builder.setTitle("Histori Shift")
                        .setMessage("Apologize! This object is under construction! Shift duration "+shiftDurationHistori)
                        // .setIcon(R.drawable.ic_android_cat)
                        .setCancelable(false)
                        .setNegativeButton("ОК",
                                new DialogInterface.OnClickListener() {
                                    public void onClick(DialogInterface dialog, int id) {
                                        dialog.cancel();
                                    }
                                });
                AlertDialog alert = builder.create();
                alert.show();

            }
        });
    }
    public void buttonToReturnClik (View view)
    {
        Intent i = new Intent(UserShiftHistory.this, UserShiftActivity.class);
        i.putExtra(Constant.USER_NAME_EMAIL, userNameEmail);
        i.putExtra(Constant.PARENT_HIERARCHY_POSITION_USER, parentHierarchyPositionUser);
        startActivity(i);
    }
}

File: android/app/src/main/java/com/TMR24/MotionStudy/UserShiftHistory.kt
package com.TMR24.MotionStudy

import android.app.AlertDialog
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.*
import android.widget.AdapterView.OnItemClickListener
import androidx.appcompat.app.AppCompatActivity
import com.TMR24.MotionStudy.UserShiftHistory
import com.google.firebase.Timestamp
import com.google.firebase.firestore.FirebaseFirestore
import java.text.SimpleDateFormat
import java.util.*

class UserShiftHistory : AppCompatActivity() {
    private var db: FirebaseFirestore? = null
    private val TAG: String? = null
    private var userNameEmail: String? = null
    private var parentHierarchyPositionUser: String? = null
    private var nameOrganization: String? = null
    private var nameSubdivision: String? = null
    private var namePosition: String? = null
    private var idPosition: String? = null
    private var idOrganization: String? = null
    private var idSubdivision: String? = null
    private var idDocPositionUser: String? = null
    private var userСomment: String? = null
    private val textActivPositionHistori: TextView? = null
    private var buttonToReturnHistori: Button? = null
    private var listViewHistoriButton: ListView? = null
    private var adapterHistoriButton: ArrayAdapter<String>? = null
    private var listHistoriButton: MutableList<String>? = null
    private var listHistoriButtonItem: MutableList<String>? = null
    private var workShiftEndTime: Timestamp? = null
    private var workShiftStartTime: Timestamp? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_user_shift_history)
        init()
        setOnClickItemlistViewHistoriButton()
    }

    private fun init() {
        db = FirebaseFirestore.getInstance()
        //   mFunctions = FirebaseFunctions.getInstance();
        buttonToReturnHistori = findViewById(R.id.buttonToReturnHistori)
        listViewHistoriButton = findViewById(R.id.listViewHistori)
        listHistoriButton = ArrayList()
        listHistoriButtonItem = ArrayList()
        adapterHistoriButton = ArrayAdapter(this, android.R.layout.simple_list_item_1, listHistoriButton)
        listViewHistoriButton.setAdapter(adapterHistoriButton)
    }

    public override fun onStart() {
        super.onStart()
        val i = intent
        if (i != null) {   //очистили ArrayList
            listHistoriButton!!.clear()
            listHistoriButtonItem!!.clear()
            // получили строку данных с предидущего Активити
            userNameEmail = i.getStringExtra(Constant.USER_NAME_EMAIL)
            parentHierarchyPositionUser = i.getStringExtra(Constant.PARENT_HIERARCHY_POSITION_USER)
            val delimeter = ">"
            idOrganization = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(0)
            nameOrganization = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(1)
            idSubdivision = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(2)
            nameSubdivision = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(3)
            idPosition = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(4)
            namePosition = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(5)
            idDocPositionUser = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(6)
            userСomment = parentHierarchyPositionUser.split(delimeter).toTypedArray().get(7)
            //вывели на экран Должность Подразделение Организацию в которой планируем работать
            val textActivPosition = findViewById<TextView>(R.id.textActivPositionHistori)
            textActivPosition.text = "$nameOrganization > $nameSubdivision > $namePosition"
            //проверили нет ли активной смены
            db!!.collection("WorkShift")
                    .whereEqualTo("EmailPositionUser", userNameEmail)
                    .whereEqualTo("WorkShiftEnd", "false")
                    .whereEqualTo("IdDocPosition", idPosition)
                    .get()
                    .addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            for (document in task.result) {
                                //Имеется активная смена
                                Log.d(TAG, document.id + " => " + document.data)
                                val doc = document.data
                                workShiftStartTime = doc["WorkShiftStartTime"] as Timestamp?
                                val g = workShiftStartTime!!.toDate()
                                val x = workShiftStartTime!!.seconds
                                val formatForDateNow = SimpleDateFormat("dd.MM.yyyy ' time ' k:mm:ss")
                                workShiftEndTime = doc["WorkShiftEndTime"] as Timestamp?
                                val h = workShiftEndTime!!.toDate()
                                val a = workShiftEndTime!!.seconds
                                val seconds = a - x
                                // переводим секунды в дни-часы-минуты-секунды
                                // long sec = seconds % 60;
                                val minutes = seconds % 3600 / 60
                                val hours = seconds % 86400 / 3600
                                val days = seconds / 86400
                                listHistoriButton!!.add("Shift start " + formatForDateNow.format(g) + " > End of shift " + formatForDateNow.format(h))
                                listHistoriButtonItem!!.add(document.id + ">" + "Day " + days + " Hour " + hours + " Minute " + minutes)
                                adapterHistoriButton!!.notifyDataSetChanged()
                            }
                        } else {
                            //отсутствует Активная смена
                            Log.d(TAG, "Error getting documents: ", task.exception)
                        }
                    }
        }
    }

    private fun setOnClickItemlistViewHistoriButton() { //
        listViewHistoriButton!!.onItemClickListener = OnItemClickListener { parent, view, position, id ->
            val buttonHistori = listHistoriButtonItem!![position]
            val delimeter = ">"
            val idButtonHistori: String = buttonHistori.split(delimeter).toTypedArray().get(0)
            val shiftDurationHistori: String = buttonHistori.split(delimeter).toTypedArray().get(1)
            // выводим диалоговое окно
            val builder = AlertDialog.Builder(this@UserShiftHistory)
            builder.setTitle("Histori Shift")
                    .setMessage("Apologize! This object is under construction! Shift duration $shiftDurationHistori") // .setIcon(R.drawable.ic_android_cat)
                    .setCancelable(false)
                    .setNegativeButton("ОК"
                    ) { dialog, id -> dialog.cancel() }
            val alert = builder.create()
            alert.show()
        }
    }

    fun buttonToReturnClik(view: View?) {
        val i = Intent(this@UserShiftHistory, UserShiftActivity::class.java)
        i.putExtra(Constant.USER_NAME_EMAIL, userNameEmail)
        i.putExtra(Constant.PARENT_HIERARCHY_POSITION_USER, parentHierarchyPositionUser)
        startActivity(i)
    }
}